dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,491 @@
+
+
+class Boolean(Value):  # Boolean
+    """boolean True/False value"""
+
+    value: bool
+
+    def __hash__(self) -> int:
+        return hash((Boolean, self.value))
+
+
+NodeType = TypeVar("NodeType", bound=Node)  # pylint: disable=C0103
+
+
+@dataclass(eq=False)
+class Alias(Named, Generic[NodeType]):
+    """wraps node types with an alias"""
+
+    child: Node
+
+    def __hash__(self) -> int:
+        return hash((Alias, self.name))
+
+
+@dataclass(eq=False)
+class Column(Named):
+    """column used in statements"""
+
+    _table: Optional["Table"] = field(repr=False, default=None)
+
+    @property
+    def table(self) -> Optional["Table"]:
+        """
+        return the table the column was referenced from
+        """
+        return self._table
+
+    def add_table(self, table: "Table") -> "Column":
+        """
+        add a referenced table
+        """
+        if self._table is None:
+            self._table = table
+        return self
+
+    def __hash__(self) -> int:
+        return hash((Column, self.name))
+
+
+@dataclass(eq=False)
+class Wildcard(Expression):
+    """wildcard or '*' expression"""
+
+    _table: Optional["Table"] = field(repr=False, default=None)
+
+    @property
+    def table(self) -> Optional["Table"]:
+        """
+        return the table the column was referenced from
+        """
+        return self._table
+
+    def add_table(self, table: "Table") -> "Wildcard":
+        """
+        add a referenced table
+        """
+        if self._table is None:
+            self._table = table
+        return self
+
+    def __hash__(self) -> int:  # pragma: no cover
+        return id(Wildcard)
+
+
+@dataclass(eq=False)
+class Table(Named):
+    """a type for tables"""
+
+    _columns: List[Column] = field(repr=False, default_factory=list)
+
+    @property
+    def columns(self) -> List[Column]:
+        """
+        return the columns referenced from this table
+        """
+        return self._columns
+
+    def add_columns(self, *columns: Column) -> "Table":
+        """
+        add columns referenced from this table
+        """
+        for column in columns:
+            self._columns.append(column)
+            column.add_table(self)
+        return self
+
+    def __hash__(self) -> int:
+        return hash((Table, self.name))
+
+
+class JoinKind(Enum):
+    """the accepted kinds of joins"""
+
+    Inner = "INNER JOIN"  # pylint: disable=C0103
+    LeftOuter = "LEFT JOIN"  # pylint: disable=C0103
+    RightOuter = "RIGHT JOIN"  # pylint: disable=C0103
+    FullOuter = "FULL JOIN"  # pylint: disable=C0103
+
+
+@dataclass(eq=False)
+class Join(Node):
+    """a join between tables"""
+
+    kind: JoinKind
+    table: Union[Table, Alias]
+    on: Expression  # pylint: disable=C0103
+
+    def __hash__(self) -> int:
+        return hash((Join, self.kind))
+
+
+@dataclass(eq=False)
+class From(Node):
+    """a from that belongs to a select"""
+
+    table: Union[Table, Alias]
```

Comment:
Do we currently support an aliased Query here? e.g.:
```
SELECT t1.a, t2.b FROM (SELECT * FROM t) t1 JOIN t2 ON t1.c = t2.c
```

Reply:
we didn't but this felt big enough to make sure it's here from the start so I've added it